### PR TITLE
chore(config): refactor subcommands

### LINKED ISSUE #2271: Configuration Improvements
## Task summary

- Refactor the configuration package so that the Gossamer CLI subcommands only expose applicable configuration parameters.

## Specification reference

- N/A

## Other information and links

-

### PR DESCRIPTION
**Pull request by danforbes**
Refactor subcommand configuration so that Gossamer subcommands only accept application configuration parameters.

## Changes

- Refactor subcommand configuration so that Gossamer subcommands only accept application configuration parameters.

## Tests

- I ran `account_test.go`, `config_test.go`, and `flags_test.go`

## Issues

- Closes #2271

## Primary Reviewer

- @noot

### CODE REVIEW

File: cmd/gossamer/config.go [hunk 1]
```diff
@@ -465,6 +434,33 @@ func setDotGlobalConfigFromToml(tomlCfg *ctoml.Config, cfg *dot.GlobalConfig) {
 	}
 }
 
+func setDotBasepathConfig(ctx *cli.Context, tomlConfig *ctoml.Config, cfg *dot.GlobalConfig) {
+	setDotBasepathConfigFromToml(tomlConfig, cfg)
+	setDotBasepathConfigFromFlags(ctx, cfg)
+}
```
Comment (qdm12) [inline cmd/gossamer/config.go:440]: nit it would be nice to have these few functions as methods on `dot.GlobalConfig` instead.

Alternatively you should change those functions to pass in values and return values, that makes it a bit easier to read.

i.e.

```go
func setDotBasepathConfig(ctx *cli.Context, tomlConfig *ctoml.Config, cfg dot.GlobalConfig) (updatedCfg dot.GlobalConfig)
```

File: cmd/gossamer/config.go [hunk 2]
```diff
@@ -465,6 +434,33 @@ func setDotGlobalConfigFromToml(tomlCfg *ctoml.Config, cfg *dot.GlobalConfig) {
...
+
+func setDotBasepathConfigFromFlags(ctx *cli.Context, cfg *dot.GlobalConfig) {
+	// check --basepath flag and update node configuration
+	if basepath := ctx.GlobalString(BasePathFlag.Name); basepath != "" {
+		cfg.BasePath = basepath
+	}
+
+	// check if cfg.BasePath his been set, if not set to default
+	if cfg.BasePath == "" {
+		cfg.BasePath = dot.GssmrConfig().Global.BasePath
```
Comment (qdm12) [inline cmd/gossamer/config.go:460]: Is this the same default base path for all `dot.*` configs?

File: cmd/gossamer/config.go [hunk 3]
```diff
@@ -465,6 +434,33 @@ func setDotGlobalConfigFromToml(tomlCfg *ctoml.Config, cfg *dot.GlobalConfig) {
...
+	}
+}
+
+func setDotBasepathConfigFromFlags(ctx *cli.Context, cfg *dot.GlobalConfig) {
+	// check --basepath flag and update node configuration
+	if basepath := ctx.GlobalString(BasePathFlag.Name); basepath != "" {
+		cfg.BasePath = basepath
+	}
+
+	// check if cfg.BasePath his been set, if not set to default
```
Comment (qdm12) [inline cmd/gossamer/config.go:458]: no need I think, the code is quite explicit

```suggestion
```

File: cmd/gossamer/main_test.go [hunk 4]
```diff
@@ -216,44 +214,6 @@ func TestInvalidCommand(t *testing.T) {
 	}
 }
 
-func TestInitCommand_RenameNodeWhenCalled(t *testing.T) {
```
Comment (qdm12) [inline cmd/gossamer/main_test.go:219]: Why was this removed?

File: cmd/gossamer/flags.go [hunk 5]
```diff
@@ -459,18 +459,19 @@ var (
...
 
 	// InitFlags are flags that are valid for use with the init subcommand
-	InitFlags = append([]cli.Flag{
+	InitFlags = []cli.Flag{
+		BasePathFlag,
 		ForceFlag,
 		GenesisFlag,
-		PruningFlag,
-		RetainBlockNumberFlag,
-	}, GlobalFlags...)
```
Comment (qdm12) [inline cmd/gossamer/flags.go:467]: Why remove the global flags 🤔 ?

**Review by qdm12 — COMMENTED** [review]
Also note the CI is failing, there might be some tests that need fixing

File: cmd/gossamer/config.go (hunk 2 above)
Comment (danforbes) [reply cmd/gossamer/config.go:459]: I don't know the answer to this question.

File: cmd/gossamer/config.go (hunk 1 above)
Comment (danforbes) [reply cmd/gossamer/config.go:440]: I think I'm going to omit that change from this PR, but I'll keep it in mind as I continue to improve the CLI stuff.

File: cmd/gossamer/config.go (hunk 2 above)
Comment (EclesioMeloJunior) [reply cmd/gossamer/config.go:459]: No, the base path is different for `gssmr` (`~/.gossamer/gssmr`), `kusama` (`~/.gossamer/kusama`) and `polkadot` (`~/.gossamer/polkadot`).

File: cmd/gossamer/flags.go (hunk 5 above)
Comment (noot) [reply cmd/gossamer/flags.go:467]: the global flags are supposed to be compatible with every subcommand, so they should be readded

File: cmd/gossamer/config.go [hunk 6]
```diff
@@ -157,20 +156,7 @@ func createInitConfig(ctx *cli.Context) (*dot.Config, error) {
...
-		return nil, err
-	}
-
-	if !cfg.Global.Pruning.IsValid() {
-		return nil, fmt.Errorf("--%s must be either %s or %s", PruningFlag.Name, pruner.Full, pruner.Archive)
-	}
-
-	if cfg.Global.RetainBlocks < dev.DefaultRetainBlocks {
-		return nil, fmt.Errorf("--%s cannot be less than %d", RetainBlockNumberFlag.Name, dev.DefaultRetainBlocks)
-	}
```
Comment (noot) [inline cmd/gossamer/config.go:173]: why is the pruning settings removed here? pruning settings need to be set upon initialization

File: cmd/gossamer/flags.go (hunk 5 above)
Comment (danforbes) [reply cmd/gossamer/flags.go:467]: Why are the global flags supposed to be compatible with every subcommand? None of the global flags apply to the init subcommand.

File: cmd/gossamer/config.go (hunk 6 above)
Comment (danforbes) [reply cmd/gossamer/config.go:173]: Ah, I did not realize that!

**Review by noot — CHANGES_REQUESTED** [review]
it looks like the pruning settings still aren't being detected properly, this is from the stress tests:
```
2022-03-03T14:47:43Z INFO loading toml configuration from /go/src/github.com/ChainSafe/gossamer/tests/utils/config_default.toml...	config.go:L48	pkg=cmd
2022-03-03T14:47:43Z WARN overwriting default configuration with id  with toml configuration values from /go/src/github.com/ChainSafe/gossamer/tests/utils/config_default.toml	config.go:L52	pkg=cmd
2022-03-03T14:47:43Z EROR failed to create node configuration: --pruning must be either full or archive	main.go:L3[13](https://github.com/ChainSafe/gossamer/runs/5408687486?check_suite_focus=true#step:5:13)	pkg=cmd
--pruning must be either full or archive
```

could you fix that / revert it back to how it was?

File: cmd/gossamer/config.go [hunk 7]
```diff
@@ -465,6 +445,47 @@ func setDotGlobalConfigFromToml(tomlCfg *ctoml.Config, cfg *dot.GlobalConfig) {
...
+func setDotBasepathConfig(ctx *cli.Context, tomlConfig *ctoml.Config, cfg *dot.GlobalConfig) {
+	setDotBasepathConfigFromToml(tomlConfig, cfg)
+	setDotBasepathConfigFromFlags(ctx, cfg)
+}
+
+func setDotBasepathConfigFromToml(tomlCfg *ctoml.Config, cfg *dot.GlobalConfig) {
+	if tomlCfg == nil {
+		return
+	}
+
```
Comment (qdm12) [inline cmd/gossamer/config.go:457]: nit do we need this check?

**Review by qdm12 — APPROVED** [review]

File: cmd/gossamer/config.go (hunk 7 above)
Comment (danforbes) [reply cmd/gossamer/config.go:457]: Based on the usage of the function, we do not need to use it at this time. However, this function was written with the goal of emulating the call to `setDotGlobalConfig`, which does have a check for a `nil` TOML config. Ref: https://github.com/ChainSafe/gossamer/blob/v0.6.0/cmd/gossamer/config.go#L421

File: cmd/gossamer/flags.go (hunk 5 above)
Comment (noot) [reply cmd/gossamer/flags.go:467]: I think all of the global flags apply to the init subcommand? which ones don't apply? log level flags should be usable by every command, same with setting node name, chain, basepath, profiling, and setting pruning settings

**Comment by timwu20** [comment]
@danforbes can you make this a draft PR if you don't intend to merge this.
